### DISCUSSION TITLE
2D Slices from 3D Input

### DISCUSSION THREAD

**Discussion by dannyhow12** [Project-MONAI/tutorials#745] [Q&A]
Hello,

Once again, thank you for the amazing repo and the ever so supportive MONAI community.

I have achieved success in implementing the 3D UNet with the MONAI framework, however I would like to now test it on a 2D UNet, where I would wish to input the 3D images that I have, have them sliced into 2D images, and to be trained with MONAI. 

I have come across the 2d_inference_3d_volume.ipynb, which gives a suggestion on how to perform so. However, it only uses a dummy example. I would appreciate if any suggestions are given on how I would input my 3D images and incorporate it into the function. Many thanks

Also, by referring to the 2d unet_trainingdict.py, it is able to load the 2D slices, however, how would I want to make sure that the slices that are input to the network are of the same case? Therefore aggregating all the slices of that case, then computing the Dice score

Best regards, 
Danny

**Comment by Nic-Ma** [comment]
Hi @dannyhow12 ,

Thanks for your interest here.
I think the `2d_inference_3d_volume.ipynb` tutorial is exactly for your use case.
@surajpaib Could you please help provide more real-world practice?

Thanks in advance.

**Reply by dannyhow12** [reply]
Thank you for the prompt reply. Perhaps I could insert the code here that I am trying with:

    from monai.utils import first, set_determinism
    from monai.transforms import (
        AsDiscrete,
        AsDiscreted,
        EnsureChannelFirstd,
        Compose,
        CropForegroundd,
        LoadImaged,
        Orientationd,
        RandCropByPosNegLabeld,
        SaveImaged,
        ScaleIntensityRanged,
        ScaleIntensityRangePercentilesd,
        NormalizeIntensityd,
        Spacingd,
        EnsureTyped,
        EnsureType,
        Invertd,
        RandCropByLabelClassesd
    )
    
    from monai.handlers.utils import from_engine
    from monai.networks.nets import UNet 
    from monai.networks.layers import Norm
    from monai.metrics import DiceMetric
    from monai.losses import DiceLoss
    from monai.inferers import sliding_window_inference, SlidingWindowInferer
    from monai.data import CacheDataset, DataLoader, Dataset, decollate_batch, list_data_collate
    from monai.config import print_config
    from monai.apps import download_and_extract
    import torch
    import matplotlib.pyplot as plt
    import tempfile
    import shutil
    import os
    import glob
    from typing import Callable, Any


    data_dir = "/content/drive/MyDrive/data_try/"
    images = sorted(glob.glob(os.path.join(data_dir, f"case_0000*", "imaging.nii.gz")))
    segs = sorted(glob.glob(os.path.join(data_dir, f"case_0000*", "segmentation.nii.gz")))
    
    print(len(images), len(segs))
    train_files = [{"image": image_name, "label": label_name} for image_name, label_name in zip(images[:-2], segs[:-2])]
    val_files = [{"image": image_name, "label": label_name} for image_name, label_name in zip(images[-2:], segs[-2:])]
    
    #print(train_files, "\Val:",val_files)
    #print(len(train_files), len(val_files))
    
    set_determinism(seed=0)

    # define transforms for image and segmentation
    train_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            EnsureChannelFirstd(keys=["image", "label"]),
            Orientationd(keys=["image", "label"], axcodes="RAS"), #Right anterior superior // (L,R); (A,P); (I,S)
            ScaleIntensityRanged(keys=["image"], a_min=-79, a_max=304, b_min=0.0, b_max=1.0, clip=True), #Clip = whether to clip after scaling is performed #-79 304 for kits
            CropForegroundd(keys=["image", "label"], source_key="image"),
            EnsureTyped(keys=["image", "label"]),
        ]
    )
    val_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            EnsureChannelFirstd(keys=["image", "label"]),
            Orientationd(keys=["image", "label"], axcodes="RAS"),
            ScaleIntensityRanged(keys=["image"], a_min=-79, a_max=304, b_min=0.0, b_max=1.0, clip=True), #Clip = whether to clip after scaling is performed #-79 304 for kits
            CropForegroundd(keys=["image", "label"], source_key="image"),
            EnsureTyped(keys=["image", "label"]),
        ]
    )


    # create a training data loader
    train_ds = CacheDataset(data=train_files, transform=train_transforms, cache_rate = 1.0, num_workers = 2)
    train_loader = DataLoader(train_ds, batch_size=1,shuffle=True,num_workers=2)
    
    # create a validation data loader
    val_ds = CacheDataset(data=val_files, transform=val_transforms,cache_rate = 1.0, num_workers = 2)
    val_loader = DataLoader(val_ds, batch_size=1, num_workers=2)
    
    class YourSlidingWindowInferer(SlidingWindowInferer):
        def __init__(self, spatial_dim: int = 0, *args, **kwargs):
            # Set dim to slice the volume across, for example, `0` could slide over axial slices,
            # `1` over coronal slices
            # and `2` over sagittal slices.
            self.spatial_dim = spatial_dim
    
            super().__init__(*args, **kwargs)
    
        def __call__(
            self,
            inputs: torch.Tensor,
            network: Callable[..., torch.Tensor],
            slice_axis: int = 0,
            *args: Any,
            **kwargs: Any,
        ) -> torch.Tensor:
    
            assert (
                self.spatial_dim < 3
            ), "`spatial_dim` can only be `[D, H, W]` with `0, 1, 2` respectively"
    
            # Check if roi size (eg. 2D roi) and input volume sizes (3D input) mismatch
            if len(self.roi_size) != len(inputs.shape[2:]):
    
                # If they mismatch and roi_size is 2D add another dimension to roi size
                if len(self.roi_size) == 2:
                    self.roi_size = list(self.roi_size)
                    self.roi_size.insert(self.spatial_dim, 1)
                else:
                    raise RuntimeError(
                        "Currently, only 2D `roi_size` is supported, cannot broadcast to volume. "
                    )
    
            return super().__call__(inputs, lambda x: self.network_wrapper(network, x))
        def network_wrapper(self, network, x, *args, **kwargs):
            """
            Wrapper handles cases where inference needs to be done using
            2D models over 3D volume inputs.
            """
            # If depth dim is 1 in [D, H, W] roi size, then the input is 2D and needs
            # be handled accordingly
    
            if self.roi_size[self.spatial_dim] == 1:
                #  Pass 4D input [N, C, H, W]/[N, C, D, W]/[N, C, D, H] to the model as it is 2D.
                x = x.squeeze(dim=self.spatial_dim + 2)
                out = network(x, *args, **kwargs)
                #  Unsqueeze the network output so it is [N, C, D, H, W] as expected by
                # the default SlidingWindowInferer class
                return out.unsqueeze(dim=self.spatial_dim + 2)
    
            else:
                return network(x, *args, **kwargs)
    
    # create UNet, DiceLoss and Adam optimizer
    device = torch.device("cuda:0")
    
    model = UNet(
        spatial_dims=2,
        in_channels=1,
        out_channels=3,
        channels=(16, 32, 64, 128, 256),
        strides=(2, 2, 2, 2),
        num_res_units=2,
        norm = Norm.BATCH,
    ).to(device)
    
    loss_function = DiceLoss(include_background = False, to_onehot_y = True, softmax = True)
    optimizer = torch.optim.Adam(model.parameters(), 1e-4)
    dice_metric = DiceMetric(include_background=False, reduction="mean")
    # dice_metric_batch = DiceMetric(include_background = False, reduction = "mean_batch") #splitting metric into corresponding kidney, tumor, background
    
    
    max_epochs = 10
    val_interval = 2
    
    best_metric = -1
    best_metric_epoch = -1
    
    epoch_loss_values = []
    
    metric_values = []
    
    post_pred = Compose([EnsureType(), AsDiscrete(argmax=True, to_onehot=3)])
    post_label = Compose([EnsureType(), AsDiscrete(to_onehot=3)])
    
    for epoch in range(max_epochs):          
        print("-" * 10)
        print(f"epoch {epoch + 1}/{max_epochs}")
        model.train()
        epoch_loss = 0
        step = 0
        for batch_data in train_loader:
            step += 1
            inputs, labels = (
                batch_data["image"].to(device),
                batch_data["label"].to(device),
            )
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = loss_function(outputs, labels)
            loss.backward()
            optimizer.step()
            epoch_loss += loss.item()
            print(
                f"{step}/{len(train_ds) // train_loader.batch_size}, "
                f"train_loss: {loss.item():.4f}")
        
        epoch_loss /= step
        epoch_loss_values.append(epoch_loss)
        print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}")
    
        if (epoch + 1) % val_interval == 0:
            model.eval()
            with torch.no_grad():
                for val_data in val_loader:
                    val_inputs, val_labels = (
                        val_data["image"].to(device),
                        val_data["label"].to(device),
                    )
    
                    roi_size = (96, 96) 
                    sw_batch_size = 4
    
                    axial_inferer = SlidingWindowInferer(roi_size=roi_size, sw_batch_size=sw_batch_size, cval=-1)
                    val_output = axial_inferer(val_inputs, model)
                    # val_outputs = sliding_window_inference(
                    #     val_inputs, roi_size, sw_batch_size, model)
                    val_outputs = [post_pred(i) for i in decollate_batch(val_outputs)]
                    val_labels = [post_label(i) for i in decollate_batch(val_labels)]
    
                    # compute metric for current iteration
                    dice_metric(y_pred=val_outputs, y=val_labels)
    
                # aggregate the final mean dice result
                metric = dice_metric.aggregate().item()
                # reset the status for next validation round
                dice_metric.reset()
                metric_values.append(metric)
    
                if metric > best_metric:
                    best_metric = metric
    
                    best_metric_epoch = epoch + 1
                    torch.save(model.state_dict(), os.path.join(
                        root_dir, "best_metric_model_1.pth"))
                    print("saved new best metric model")
                print(
                    f"\ncurrent epoch: {epoch + 1} current mean dice: {metric:.4f}"
                    f"\nbest mean dice: {best_metric:.4f} "
                    f"at epoch: {best_metric_epoch}"
                )

While the error that prompts when I run the training block is as below:

    RuntimeError: Expected 3D (unbatched) or 4D (batched) input to conv2d, but got input of size: [1, 1, 502, 386, 834]

With this being said, I believe that I might have used the slidingwindowinferer wrongly, thus would appreciate if any form of advice is given @surajpaib . Thank you in advance.

Best regards, 
Danny

**Reply by Nic-Ma** [reply]
Hi @dannyhow12 ,

I think maybe you should use the `SliceInferer` inferer:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/inferers/inferer.py#L256
It's the same as the idea in the tutorial as @surajpaib showed with a customized inferer.

Thanks.

**Reply by dannyhow12** [reply]
Hi @Nic-Ma ,

Yep, I have changed to that, however the line that is prompting the error is

---> 46         outputs = model(inputs)

Do I need to slice the images before hand, or is it possible to apply any monai transforms that is able to slice 3D volumes into its corresponding slices?

**Reply by Nic-Ma** [reply]
Hi @surajpaib ,

I don't remember clearly about it, could you please help share some best practices?

Thanks in advance.

**Reply by dannyhow12** [reply]
Hi @Nic-Ma @surajpaib ,

With reference to the problem as mentioned by https://github.com/Project-MONAI/MONAI/discussions/3901, the SqueezeDimd transforms can be added. However, I am unsure of what the argument dim does in this transform, thus I only set it to 0, which is the default value.

I did not change much of the other codes, however this is what I noticed when I tried to run it:

         46         optimizer.zero_grad()
         47         print(inputs.shape, labels.shape)
    ---> 48         outputs = model(inputs)
         49         loss = loss_function(outputs, labels)
         50         loss.backward()

The error

    RuntimeError: Given groups=1, weight of size [16, 1, 3, 3], expected input[1, 512, 337, 611] to have 1 channels, but got 512 channels instead

I then proceed to print its image shape and label shape,  where it shows 

    torch.Size([1, 512, 337, 611]) torch.Size([1, 512, 337, 611])

However, it is supposed to be 1,512,512,611, I am unsure why it changed to an arbitrary number of 337. I am thinking it is due to the value of dim in SqueezeDim, however I am unable to find any reliable references to workaround. Appreciate it alot if any form of advices are given.

Best regards,
Danny

**Reply by surajpaib** [reply]
Hi @dannyhow12 

The `SliceInferer` is to be used for evaluation purposes only (val/test) as it contains multiple forward pass calls. I can see from the code you shared above that you are indeed using it only for validation. 

If I understand correctly, your issue above seems to be getting 2D slices generated from a 3D volume during training. You will need a transform to go from 3D to 2D. It seems like you're using the SqueezeDim without actually generating a 2D slice from the 3D which will probably just squeeze one of the dims from your 3D (NCDHW) tensor (in your case it seems like its squeezing the 0th dimension). So you'll probably need to use a transform to generate the 2D slice before using Squeezedim. I'm not an expert on monai transform structure so I'll direct this back to @Nic-Ma 

P.S: for the issue of `337`, I think it might be due to the `CropForegroundd(keys=["image", "label"], source_key="image")` transform.

**Reply by dannyhow12** [reply]
Hi @surajpaib ,

Thanks for pointing out on issue of `337`, it was solved after removing the cropforegroundd transform. I see, I thought that SqueezeDimd transform was already performing slicing of a 3D volume, would appreciate if @Nic-Ma could share more on the available transform while I work on this issue.

Thank you in advance

**Reply by Nic-Ma** [reply]
@surajpaib thanks very much for your help here!

And about the `SeqeezeDim` question, I think you can refer to: https://github.com/Project-MONAI/MONAI/issues/299#issuecomment-618913381
```py
train_transforms = Compose([
    LoadImaged(keys=['img', 'seg']),
    EnsureChannelFirstd(keys=['img', 'seg']),
    NormalizeIntensityd(keys=['img']),
    RandSpatialCropd(keys=['img', 'seg'], roi_size=[96, 96, 1], random_size=False),
    SqueezeDimd(keys=['img', 'seg'], dim=-1),
    ToTensord(keys=['img', 'seg'])
])
```
@martaranzini helped a lot with this use case.
@martaranzini please feel free to correct me if I misunderstood anything.

Thanks.

**Reply by dannyhow12** [reply]
Hi @Nic-Ma , 

Thank you! However with the RandSpatialCropd transform, it essentially only crops 1 sample per image (do correct me if I am wrong), thus it would not be so applicable in my case where I would want to decompose a 3D CT image into its corresponding slices. With this being said, why isnt the RandSpatialCropSamplesd transform used instead in this case, as the number of samples generated could be defined. 

Thank you in advance

**Reply by wyli** [reply]
there are two ways of getting image patches from a volume for training, please see: https://github.com/Project-MONAI/MONAI/discussions/4196#discussioncomment-2669039 difference is in whether the slices from different volumes should be shuffled by the training data loader - the first method doesn't achive the shuffling.

**Reply by Nic-Ma** [reply]
Hi @dannyhow12 ,

Thanks for @wyli 's quick update, he put together the typical reference examples: https://github.com/Project-MONAI/tutorials/pull/747
Please take a look.

Thaks.

**Reply by dannyhow12** [reply]
Hi @wyli @Nic-Ma ,

Thank you so much for the quick updates! I am eternally grateful, I will try it out and revert back if there are any positive or negative advancements regarding this method. 

Best regards,
Danny

**Reply by dannyhow12** [reply]
Hi @wyli @Nic-Ma , hope you all have a great week ahead

I have tried out the section of the code to incorporate it to my dataset. The problem that I face is: 

Prerequisite: 
My datasets generally are of the size 512 x 512 x z, where z is the varying number of slices in a 3D CT Scan. 

1) After applying the first train_transforms, the check_loader returns a shape of [1, 512, 512, 611], which is applicable as the channel is set to be first with AsChannelFirstd, and the last slice is where we want to perform our slicing. 
2) At the second data_loader, where it incorporates patch_func and patch_transform. However, when the check_loader returns the shape, it outputs an error of 

        File "/usr/local/lib/python3.7/dist-packages/monai/utils/misc.py", line 154, in ensure_tuple_rep
        raise ValueError(f"Sequence must have length {dim}, got {len(tup)}.")
        ValueError: Sequence must have length 2, got 3.

By going through the patch_func and patch_transform, it should have already sliced to a 2d patch of 48 by 48. Could it be due to the problem of not defining the axis of my dataset properly?

Best regards,
Danny

**Reply by wyli** [reply]
you are right, `AsChannelFirstd` is not robust, I should have used `EnsureChannelFirstd`. could you please try the same modification as this commit: https://github.com/Project-MONAI/tutorials/pull/747/commits/b4648a5d6093b930dc5cdb0343b13f0707242463  you should get the output shape for your datasets:
```
first volume's shape:  torch.Size([1, 1, 512, 512, z]) torch.Size([1, 1, 512, 512, z])
first patch's shape:  torch.Size([3, 1, 48, 48]) torch.Size([3, 1, 48, 48])
```

**Reply by dannyhow12** [reply]
Hi @wyli @Nic-Ma ,

1) I do not really understand the 3D transform of RandCropbyPosNegLabeld. 4 samples are generated from each image, with a spatial size of [-1, -1. 1]. How does this achieve slicing all z slices of a specified image? If I understand correctly, it only slices 4 images with a spatial size of 512, 512, 1 (referring to my case where all images are D x H x W of z x 512 x 512). Or is this achieved by the SqueezeDimd? Thank you for your guidance.
2) As each image will be sliced to its corresponding number of slices, thus I would want to use CacheDataset to speed up the training, as in the spleen segmentation notebook example. However, each time when I run the training with CacheDataset caching my image with its transforms, the RAM overwhelms and crashes (Note: I am using Google Colab to train, with a RAM of 12GB). The conventional Dataset works fine, but it is much slower. 

Thank you in advance.

Best regards,
Danny

**Reply by wyli** [reply]
Hi @dannyhow12, the previous tutorial was not trying to address your initial ask. Now I have updated to include another notebook to fetch all z slices which is what you want. However because the number of slices is different across the volumes, we can only use an iterable dataset instead of a regular dataset with known total slices.

https://github.com/Project-MONAI/tutorials/blob/1dfa20ca373a582f404856b3ff509958655b183f/modules/2d_slices_from_3d_training.ipynb

(this notebook requires monai v0.9.0, it'll be officially released early next week.)
